### ISSUE TITLE
CA1062 should support fluent syntax

### ISSUE THREAD

**Issue by thunderbird55** [dotnet/roslyn-analyzers#3340]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

CA1062

#### Repro steps
```csharp        
public static void Foo(string bar)
{
       Guard.Argument(bar, nameof(bar)).NotNull();
       var length = bar.Length; // Warning CA1062
}
```

#### Expected behavior

Add configuration options to support fluent syntaxes.
Maybe something like dotnet_code_quality.null_check_validation_methods = Argument->NotNull|...

#### Actual behavior

Only 'usual' methods that check directly the value can be marked as null_check_validation_methods.

**Comment by Evangelink** [comment]
Are you referring to [safakgur/guard](https://github.com/safakgur/guard)?

That would make sense to allow user to define methods validating null but I am not sure how this would help in this case because `Guard.NotNull` is the method doing the validation but at this point we don't know what's the parameter.

@sharwell I have looked at the various attributes available but I can't manage to find one that would help in the context of this method definition: https://github.com/safakgur/guard/blob/7080fcad4dffc5e07b3f166e869ebe8cf111ed2a/src/Guard.Null.cs#L90-L103
Do you have any idea? If not, do you recommend opening a ticket on roslyn or on csharplang?

**Comment by mavasani** [comment]
@Evangelink I believe the proposal from @thunderbird55 is that `Argument` will be considered as a Null validation method for its argument(s), only if it's return value is an invocation receiver for `Null` method. This seems like a reasonable feature request.

**Comment by Evangelink** [comment]
Ok so we are ignoring chain calls where the `NotNull` is not right after the `Argument`?

**Comment by thunderbird55** [comment]
I use my own implementation of a fluent syntax, but its basically the same as https://github.com/safakgur/guard

I didn't consider a solution with attributes and I think this would be a better solution because the code itself tells that it validates. Maybe something like this.
```csharp
public static class Guard
{
	[return: ValidatorObject(nameof(value)] // Specify the name of the argument that is hold for validation
	public static ArgumentInfo<T> Argument<T>(T value, string name)
	{
		return new ArgumentInfo<T>(value, name);
	}
	
	[NotNullValidation]
	public static ArgumentInfo<T> NotNull(this ArgumentInfo<T> argumentInfo)
	{
	  // ...
	}
}
```
@mavasani Yes that was my first intention. It would be nice if chain calls would be supported but usually  a null check is the first validation, so this is no must-have.

**Comment by Evangelink** [comment]
@mavasani let me know if you want me to handle this by adding an exclusion for a `NotNull` invocation call on an `ArgumentInfo<T>` object. I think that we will need to exclude the namespaces as @thunderbird55 implementation probably has different namespace than the safakgur's one.

I still think that having attributes would be better but that would required more effort.

**Comment by thunderbird55** [comment]
Hi, are there any news on this? 
We would like to use the fluent syntax, but this is not possible until this is solved because we have a zero-warning policy.
Best regards

**Comment by sharwell** [comment]
It seems like the fluent syntax and CA1062 were meant for two non-overlapping sets of users. My recommendation would be for users to disable CA1062 and implement a new set of analyzers specific to their validation scenarios, which in addition to recognizing the fluent syntax could warn for use of _non_-fluent syntax.

**Comment by Evangelink** [comment]
@mavasani Could we close this ticket and suggest that the fluent validation frameworks provide a suppressor for this rule instead? 

I have been looking at a few solutions and I can't find any elegant generic way in which we could handle this. The best solution I could came up with is something along the lines of "method signature with brackets around the argument doing the capture" + "->" + "null check method signature". In this case that would be something like: `M:Guard.Argument([T],string)->M:Guard.NotNull(ArgumentInfo)`

**Comment by mavasani** [comment]
> Could we close this ticket and suggest that the fluent validation frameworks provide a suppressor for this rule instead?

That seems reasonable to me. @sharwell, if you agree, we can close this issue as by design.

**Comment by sharwell** [comment]
That seems reasonable to me
